File: timebridge/timebridge/report/attendance_report/attendance_report.js
// Copyright (c) 2026, UPGO and contributors
// For license information, please see license.txt

frappe.query_reports["Attendance Report"] = {

    filters: [
        // Always one calendar month. A register is printed and signed per
        // month, so a free date range would produce a sheet nobody files.
        {
            fieldname: "month",
            label: __("Month"),
            fieldtype: "Select",
            default: (new Date().getMonth() + 1).toString(),
            options: [
                { value: "1", label: __("January") },
                { value: "2", label: __("February") },
                { value: "3", label: __("March") },
                { value: "4", label: __("April") },
                { value: "5", label: __("May") },
                { value: "6", label: __("June") },
                { value: "7", label: __("July") },
                { value: "8", label: __("August") },
                { value: "9", label: __("September") },
                { value: "10", label: __("October") },
                { value: "11", label: __("November") },
                { value: "12", label: __("December") }
            ]
        },
        {
            fieldname: "year",
            label: __("Year"),
            // A number box rather than a list: the arrows step through years
            // without opening anything, and a year outside a fixed list can
            // still be typed straight in.
            fieldtype: "Int",
            default: new Date().getFullYear()
        },

        {
            fieldname: "organization",
            label: __("Organization"),
            fieldtype: "Link",
            options: "Organization"
        },
        {
            fieldname: "branch",
            label: __("Branch"),
            fieldtype: "Link",
            options: "Branch",
            get_query: function () {
                const organization = frappe.query_report.get_filter_value("organization");
                return organization ? { filters: { organization: organization } } : {};
            }
        },
        {
            fieldname: "department",
            label: __("Department"),
            fieldtype: "Link",
            options: "Department",
            get_query: function () {
                const branch = frappe.query_report.get_filter_value("branch");
                return branch ? { filters: { branch: branch } } : {};
            }
        },
        {
            fieldname: "shift",
            label: __("Shift"),
            fieldtype: "Link",
            options: "Shift"
        },
        {
            fieldname: "employee",
            label: __("Employee"),
            fieldtype: "Link",
            options: "Employee"
        },
        {
            fieldname: "include_inactive",
            label: __("Include Inactive"),
            fieldtype: "Check",
            default: 0
        }
    ],

    onload: function (report) {
        add_export_buttons(report, { landscape: true });
        add_year_steppers(report);
    },

    get_datatable_options: function (options) {
        // The row-number column ships without a heading, which leaves an
        // unlabelled column of digits beside the codes.
        return Object.assign(options, { serialNoColumn: true, checkboxColumn: false });
    },

    after_datatable_render: function (datatable) {

        // Filters can be rebuilt from under us, taking the arrows with them.
        // Re-attaching after every render is cheaper than tracking that;
        // add_year_steppers does nothing when they are already in place.
        add_year_steppers(frappe.query_report);

        // Shade the date itself on a weekly off, the way a paper register
        // does. The column definition carries the flag; the header cell is
        // not reachable from the formatter, which only ever sees body cells.
        shade_weekly_off_headers(datatable);
    },

    formatter: function (value, row, column, data, default_formatter) {

        // One letter per day, colour-coded.
        if (column.fieldname && column.fieldname.startsWith("day_")) {

            // Colours chosen to survive a black-and-white printer: absence is
            // the only one that stays visibly darker, because it is the one
            // people scan the page looking for.
            const colours = {
                P: "var(--green-600)",
                A: "var(--red-500)",
                "P/A": "var(--orange-500)",
                S: "var(--blue-600)",
                L: "var(--blue-500)",
                H: "var(--purple-600)",
                R: "var(--gray-600)",
                I: "var(--purple-500)",
                "?": "var(--orange-400)"
            };

            const colour = colours[value] || "var(--text-muted)";
            const weight = value === "A" ? "700" : "500";

            return `<div style="text-align:center;color:${colour};font-weight:${weight}">${value || ""}</div>`;
        }

        value = default_formatter(value, row, column, data);

        // Two destinations, because they answer different questions. The name
        // opens the day-by-day detail — the times behind these letters, for
        // the month already on screen. The code opens the Employee record
        // itself, for the shift, phone number and device mapping.
        //
        // The row carries the employee id; the name alone could not be used,
        // because employees are named EMP-##### and two people may share a
        // name.
        //
        // Neither is styled as a link. This sheet is printed, and a column of
        // blue underlined names would carry that ink onto paper for no reason.
        // They reveal themselves on hover instead.
        if (data && data.employee) {

            // Hand the name over before anyone clicks.
            //
            // Employee is shown by title rather than id, so the detail report's
            // Link filter would otherwise have to fetch that title from the
            // server — and it does so without waiting, leaving the filter box
            // empty for a moment. The report reads its filters out of the DOM,
            // finds nothing, and renders "pick an employee"; only a manual
            // refresh then worked.
            //
            // This row already holds both halves, and navigation is in-page, so
            // seeding the cache here means the filter fills in synchronously
            // and there is no moment to lose the race in.
            frappe.utils.add_link_title("Employee", data.employee, data.employee_name);

            if (column.fieldname === "employee_name") {

                value = `<a href="${detail_route(data.employee)}"
                            class="tb-employee-link"
                            title="${__("Open day-by-day detail")}">${value}</a>`;

            } else if (column.fieldname === "employee_code") {

                value = `<a href="/app/employee/${encodeURIComponent(data.employee)}"
                            class="tb-employee-link"
                            title="${__("Open employee record")}">${value}</a>`;
            }
        }

        // Absence is the one total worth spotting without reading across.
        if (column.fieldname === "total_absent" && data && data.total_absent > 0) {
            value = `<span style="color: var(--text-on-red, #b34d4d);">${value}</span>`;
        }

        return value;
    }

};


/**
 * Where a name click lands: the same month, for one person, with the times.
 *
 * The period travels in the URL because Frappe turns query parameters into
 * filters (router.js, set_route_options_from_url). Without it the detail would
 * open on the current month while the register behind it showed another one,
 * and the two would quietly disagree.
 */
function detail_route(employee) {

    const filters = frappe.query_report.get_filter_values() || {};

    const params = new URLSearchParams({
        employee: employee,
        month: filters.month || (new Date().getMonth() + 1).toString(),
        year: filters.year || new Date().getFullYear()
    });

    return `/app/query-report/Employee Attendance Detail?${params.toString()}`;
}


/**
 * Put Excel, PDF and Print on the toolbar instead of inside the "..." menu.
 *
 * Frappe's own methods are called rather than reimplemented, so exports keep
 * working the way the menu does — same filters, same access log, same naming.
 */
function add_export_buttons(report, opts) {

    opts = opts || {};

    // Third argument is the group name, which becomes the dropdown's label.
    // Deliberately not "Actions": that is Frappe's own menu for chart and card
    // tools, and mixing exports into it buries three things people use daily
    // among three they rarely touch. Their own dropdown sits beside it.
    const EXPORT = __("Export");

    report.page.add_inner_button(__("Excel"), function () {
        // Frappe's export dialog already defaults to Excel: one click, then
        // Download.
        report.export_report();
    }, EXPORT);

    report.page.add_inner_button(__("PDF"), function () {
        open_print_settings(report, opts, function (print_settings) {
            report.pdf_report(print_settings);
        });
    }, EXPORT);

    report.page.add_inner_button(__("Print"), function () {
        open_print_settings(report, opts, function (print_settings) {
            report.print_report(print_settings);
        });
    }, EXPORT);
}


function open_print_settings(report, opts, run) {

    const dialog = frappe.ui.get_print_settings(
        false,
        run,
        report.report_doc.letter_head,
        report.get_visible_columns()
    );

    // Thirty-eight columns never fit portrait A4. Preselecting landscape saves
    // discovering that by wasting a sheet.
    if (opts.landscape && dialog.fields_dict.orientation) {
        dialog.set_value("orientation", "Landscape");
    }

    report.add_portrait_warning(dialog);
}


/**
 * Colour the date headers that fall on a weekly off.
 *
 * The report's formatter is only given body cells, so the header has to be
 * reached through the rendered table. Which columns to shade comes from the
 * `weekly_off` flag the server puts on each day column — the same flag the
 * attendance calculation uses, so the register cannot disagree with the
 * records it is printing.
 */
function shade_weekly_off_headers(datatable) {

    if (!datatable || !datatable.datamanager) {
        return;
    }

    const columns = datatable.datamanager.columns || [];

    columns.forEach((col) => {

        if (!col.weekly_off) {
            return;
        }

        // Header cells carry their own class. `dt-cell--col-N` also matches
        // the inline filter row below the headings, which is why the shading
        // was landing on a row of empty search boxes instead of the date.
        const $cell = $(datatable.header).find(`.dt-cell--header-${col.colIndex}`);

        $cell.css({
            "background-color": "#ffd7d7",
            "font-weight": "700"
        });
    });
}


/**
 * Replace the year box's spinner with two arrows that step the year.
 *
 * A dropdown meant opening a list to move one year — three actions for a
 * one-step change. The browser's own spinner runs the wrong way round for a
 * calendar: here the up arrow goes back in time, not forward.
 *
 * The arrows are hung off the input's own parent. Filters in the page header
 * are built in "only_input" mode, which has no .control-input div at all — an
 * earlier attempt targeted that and silently attached nothing.
 */
function add_year_steppers(report) {

    const filter = report.get_filter && report.get_filter("year");

    if (!filter || !filter.$input || filter.$wrapper.find(".tb-year-steps").length) {
        return;
    }

    const $input = filter.$input;
    const $holder = $input.parent();

    // Frappe's Int control is an <input type="text">, so there is no native
    // spinner to hide — every arrow in this box is one of ours.
    //
    // The left side keeps Frappe's own 8px so the year still lines up with the
    // month beside it. Only the right is widened, far enough that the digits
    // never reach the arrows.
    $input.css("padding-right", "34px");
    $holder.css("position", "relative");

    $holder.append(`
        <div class="tb-year-steps" style="position:absolute;right:18px;top:50%;
             transform:translateY(-50%);display:flex;flex-direction:column;
             line-height:7px;cursor:pointer;user-select:none;
             color:var(--text-muted);z-index:2">
            <span class="tb-year-prev" title="${__("Previous year")}"
                  style="font-size:8px;padding:1px 2px">&#9650;</span>
            <span class="tb-year-next" title="${__("Next year")}"
                  style="font-size:8px;padding:1px 2px">&#9660;</span>
        </div>
    `);

    const step = (by) => {
        const now = cint(filter.get_value()) || new Date().getFullYear();
        filter.set_value(now + by);
        report.refresh();
    };

    // Up goes to the earlier year, down to the later one — as asked for.
    $holder.find(".tb-year-prev").on("click", () => step(-1));
    $holder.find(".tb-year-next").on("click", () => step(1));
}


frappe.dom.set_style(`
    /* Reads as ordinary text until pointed at, so the printed register keeps
       looking like a register. */
    a.tb-employee-link,
    a.tb-employee-link:visited {
        color: inherit;
        text-decoration: none;
    }
    a.tb-employee-link:hover {
        color: var(--primary);
        text-decoration: underline;
    }
    @media print {
        a.tb-employee-link { color: inherit !important; text-decoration: none !important; }
    }
`);

File: timebridge/timebridge/report/attendance_report/attendance_report.py
# Copyright (c) 2026, UPGO and contributors
# For license information, please see license.txt

"""
The monthly attendance register: one row per employee, one column per day.

A totals-only Summary view lived here too, behind a View filter. It has been
removed — the register already shows every day and carries its own totals, so
the second view was a different arrangement of the same numbers with a filter
to maintain and a choice to make before either could be read.

The period is always a calendar month rather than a free date range, because
this sheet gets printed and signed one month at a time.
"""

import calendar

import frappe

from frappe.utils import cint, get_last_day, getdate

# One letter per day, so 31 columns still fit across a printed page.
STATUS_CODE = {
    "Present": "P",
    "Absent": "A",
    # Half a day is written the way a paper register writes it — half present,
    # half absent — rather than as a letter someone has to look up.
    "Half Day": "P/A",
    "On Leave": "L",
    "Holiday": "H",
    "In Office": "I",
    "Needs Review": "?",
}

# Sick leave is shown apart from other leave: it is the one with a paid quota,
# so counting it in with the rest would hide the thing worth watching.
SICK_LEAVE_CODE = "S"

WEEKLY_OFF_CODE = "R"
NO_RECORD_CODE = ""

LEGEND = (
    "P = Present &nbsp; A = Absent &nbsp; P/A = Half Day &nbsp; S = Sick Leave &nbsp; L = Other Leave &nbsp; "
    # "I = In Office" is left out on purpose. It can only ever appear against
    # today — someone who has punched in and not yet out — so on a register for
    # a finished month it is a line of legend explaining a letter that is not
    # on the page.
    "H = Holiday &nbsp; R = Rest (Weekly Off) &nbsp; ? = Needs Review"
)


def execute(filters=None):

    return day_wise(frappe._dict(filters or {}))


# ---------------------------------------------------------------- shared

def get_employees(filters):

    conditions = ""
    values = {}

    for field in ("organization", "branch", "department", "shift"):

        if filters.get(field):
            conditions += f" AND emp.{field} = %({field})s"
            values[field] = filters.get(field)

    if filters.get("biometric_machine"):

        # Membership is read from the Machine User links rather than from
        # Employee.biometric_machine, because the link is what attendance
        # actually follows, and that field can only name one machine even for
        # somebody enrolled on two.
        conditions += """
            AND emp.name IN (
                SELECT mu.employee
                FROM `tabMachine User` mu
                WHERE mu.machine = %(biometric_machine)s
                  AND mu.employee IS NOT NULL
            )
        """
        values["biometric_machine"] = filters.get("biometric_machine")

    if filters.get("employee"):
        conditions += " AND emp.name = %(employee)s"
        values["employee"] = filters.get("employee")

    if not filters.get("include_inactive"):
        conditions += " AND emp.is_active = 1"

    return frappe.db.sql(
        f"""
        SELECT
            emp.name, emp.employee_code, emp.employee_name,
            emp.organization, emp.branch, emp.department, emp.shift
        FROM `tabEmployee` emp
        WHERE 1 = 1 {conditions}
        ORDER BY emp.employee_name ASC
        """,
        values,
        as_dict=True,
    )


# --------------------------------------------------------------- day-wise

def day_wise(filters):

    month, year = resolve_period(filters)
    days = calendar.monthrange(year, month)[1]

    off_days = weekly_off_dates(year, month, days)
    employees = get_employees(filters)

    columns = day_columns(year, month, days, off_days)
    heading = day_wise_heading(filters, month, year)

    if not employees:
        return columns, [], heading

    records = day_records(employees, year, month)

    # Ordered by code, read as a number. A plain text sort puts 10 before 2 and
    # 1000 in the middle, which looks like a mistake on a printed register.
    employees = sorted(employees, key=code_sort_key)

    return columns, day_rows(employees, records, off_days, days, month, year), heading


def code_sort_key(employee):

    code = (employee.employee_code or "").strip()

    # Codes are numeric here, but a lettered one (F9, F10) must not crash the
    # sort — it simply goes to the end, in its own alphabetical order.
    if code.isdigit():
        return (0, int(code), "")

    return (1, 0, code.lower())


def day_wise_heading(filters, month, year):
    """
    The title block printed above the table.

    Frappe renders this on screen and carries it into Print and PDF, which is
    where a sheet needs to identify itself — one printed page of letters is
    meaningless without knowing whose staff and which month.

    It does not reach the Excel export; Frappe writes only the grid there. Tick
    "Include filters" when exporting if the period needs to travel with it.
    """

    organisation = filters.get("organization") or frappe.db.get_value(
        "Organization", {}, "name"
    )

    name = frappe.db.get_value("Organization", organisation, "organization_name") or ""

    period = f"{calendar.month_name[month]} {year}"

    # The title block is for paper. On screen the organisation and month are
    # already sitting in the filter bar a few pixels above, so repeating them
    # only pushes the register itself further down. Hidden on screen, shown
    # when printed — the legend stays visible in both, since the letters are
    # meaningless without it.
    return (
        "<style>"
        "  .tb-print-title { display: none; }"
        "  @media print { .tb-print-title { display: block !important; } }"
        "</style>"
        "<div class='tb-print-title' style='text-align:center;line-height:1.5;margin-bottom:6px'>"
        f"<div style='font-size:15px;font-weight:700'>{frappe.utils.escape_html(name)}</div>"
        f"<div style='font-weight:600'>Attendance Register for {period}</div>"
        "</div>"
        f"<div style='text-align:center;font-size:11px;color:#777'>{LEGEND}</div>"
    )


def resolve_period(filters):

    if filters.get("month") and filters.get("year"):
        return cint(filters.month), cint(filters.year)

    today = getdate()

    return today.month, today.year


def day_columns(year, month, days, off_days):

    columns = [
        # No serial column of our own: Frappe numbers every row already, and a
        # second count beside it looks like a mistake — it was also being
        # summed into a meaningless total at the foot of the page.
        {"label": "Code", "fieldname": "employee_code", "fieldtype": "Data", "width": 70},
        {"label": "Full Name Of The Employee", "fieldname": "employee_name",
         "fieldtype": "Data", "width": 190},
    ]

    for day in range(1, days + 1):

        columns.append({
            # Just the number. "1 W" ran together into something unreadable,
            # and the weekly off is now shown by shading the header instead.
            "label": str(day),
            "fieldname": f"day_{day}",
            "fieldtype": "Data",
            # 38px could not fit a two-digit date beside the sort marker, so
            # every day past the 9th showed as "1...".
            "width": 50,
            # Read by the client formatter to shade the whole column, so a
            # weekly off is obvious at a glance instead of having to count
            # across thirty-one narrow boxes.
            "weekly_off": 1 if day in off_days else 0,
        })

    columns += [
        {"label": "Total Month Day", "fieldname": "month_days", "fieldtype": "Int", "width": 90},
        {"label": "P", "fieldname": "total_present", "fieldtype": "Int", "width": 45},
        {"label": "P/A", "fieldname": "total_half", "fieldtype": "Int", "width": 55},
        {"label": "L", "fieldname": "total_leave", "fieldtype": "Int", "width": 45},
        {"label": "A", "fieldname": "total_absent", "fieldtype": "Int", "width": 45},
    ]

    return columns


def day_records(employees, year, month):
    """Attendance for the month, keyed by (employee, day-of-month)."""

    first = getdate(f"{year}-{month:02d}-01")
    last = get_last_day(first)

    rows = frappe.db.sql(
        """
        SELECT att.employee, DAY(att.attendance_date) AS day, att.status,
               att.total_hours, att.leave_type, lv.is_paid, lv.half_day
        FROM `tabTimeBridge Attendance` att
        LEFT JOIN `tabTimeBridge Leave` lv
               ON lv.employee = att.employee
              AND lv.approval_status = 'Approved'
              AND att.attendance_date BETWEEN lv.from_date AND lv.to_date
        WHERE att.employee IN %(employees)s
          AND att.attendance_date BETWEEN %(first)s AND %(last)s
        """,
        {"employees": [e.name for e in employees], "first": first, "last": last},
        as_dict=True,
    )

    return {(r.employee, r.day): r for r in rows}


def weekly_off_dates(year, month, days):
    """
    Which days of this month are weekly offs.

    Read from the same setting attendance uses, so the register cannot
    disagree with the records it is printing.
    """

    from timebridge.timebridge.services.attendance_sync import weekly_off_days

    off_weekdays = weekly_off_days()

    return {
        day
        for day in range(1, days + 1)
        if calendar.weekday(year, month, day) in off_weekdays
    }


def day_rows(employees, records, off_days, days, month, year):

    data = []

    for emp in employees:

        row = {
            "employee": emp.name,
            "employee_code": emp.employee_code,
            "employee_name": emp.employee_name,
            "month_days": days,
        }

        present = absent = leave = half = 0

        for day in range(1, days + 1):

            record = records.get((emp.name, day))

            if not record:
                # A weekly off with no record is still a weekly off; anything
                # else genuinely has no information and is left blank rather
                # than guessed at.
                row[f"day_{day}"] = WEEKLY_OFF_CODE if day in off_days else NO_RECORD_CODE
                continue

            code = STATUS_CODE.get(record.status, "?")

            if record.status == "On Leave" and (record.leave_type or "") == "Sick Leave":
                code = SICK_LEAVE_CODE

            row[f"day_{day}"] = code

            if record.status == "Present":
                present += 1

            elif record.status == "Half Day":
                half += 1

            elif record.status == "Absent":
                absent += 1

            elif record.status == "On Leave":
                leave += 1

        row["total_present"] = present
        row["total_half"] = half
        row["total_absent"] = absent
        row["total_leave"] = leave

        data.append(row)

    return data

